### PR TITLE
add alltoall api

### PR DESCRIPTION
**Pull request by sandyhouse**
### PR types

New features
### PR changes

APIs
### Describe

Add paddle.distributed.alltoall api

How to use:
```python
            import numpy as np
            import paddle
            from paddle.distributed import init_parallel_env
            init_parallel_env()
            out_tensor_list = []
            if paddle.distributed.ParallelEnv().rank == 0:
                np_data1 = np.array([[1, 2, 3], [4, 5, 6]])
                np_data2 = np.array([[7, 8, 9], [10, 11, 12]])
            else:
                np_data1 = np.array([[13, 14, 15], [16, 17, 18]])
                np_data2 = np.array([[19, 20, 21], [22, 23, 24]])
            data1 = paddle.to_tensor(np_data1)
            data2 = paddle.to_tensor(np_data2)
            paddle.distributed.alltoall([data1, data2], out_tensor_list)
```

### CODE REVIEW

File: python/paddle/distributed/collective.py [hunk 1]
```diff
@@ -954,3 +955,73 @@ def split(x,
...
+
+    ring_id = 0 if group is None else group.id
+    op_type = 'alltoall'
+    temp = paddle.concat(in_tensor_list, axis=0)
+    helper = LayerHelper(op_type, **locals())
+    nranks = len(in_tensor_list)
+    out = helper.create_variable_for_type_inference(
+        dtype=in_tensor_list[0].dtype)
+    if in_dygraph_mode():
+        core.ops.alltoall_(temp, out, 'use_calc_stream', use_calc_stream,
```
Comment (ForFishes) [inline python/paddle/distributed/collective.py]: need rm ``out`` in inplace strategy, and fix ``op_function_generator.cc ``

**Review by ForFishes — DISMISSED** [review]
LGTM

File: python/paddle/distributed/collective.py [hunk 2]
```diff
@@ -954,3 +955,73 @@ def split(x,
...
+            out_tensor_list = []
+            if paddle.distributed.ParallelEnv().rank == 0:
+                np_data1 = np.array([[1, 2, 3], [4, 5, 6]])
+                np_data2 = np.array([[7, 8, 9], [10, 11, 12]])
+            else:
+                np_data1 = np.array([[13, 14, 15], [16, 17, 18]])
+                np_data2 = np.array([[19, 20, 21], [22, 23, 24]])
+            data1 = paddle.to_tensor(np_data1)
+            data2 = paddle.to_tensor(np_data2)
+            paddle.distributed.all_to_all([data1, data2], out_tensor_list)
```
Comment (wangxicoding) [inline python/paddle/distributed/collective.py:1206]: 可以把跑完后的结果也放到文档里

File: python/paddle/distributed/collective.py [hunk 3]
```diff
@@ -954,3 +955,73 @@ def split(x,
...
+
+def alltoall(in_tensor_list, out_tensor_list, group=None, use_calc_stream=True):
+    """
+    Scatter tensors in in_tensor_list to all participators and gather the result tensors in out_tensor_list.
+    Args:
+        in_tensor_list (list): A list of input Tensors. Every element in the list must be a Tensor whose data type
+            should be float16, float32, float64, int32 or int64.
+        out_tensor_list (Tensor): A list of output Tensors. The data type of its elements should be the same as the
+            data type of the input Tensors.
+        group (Group): The group instance return by new_group or None for global default group.
```
Comment (TCChenlong) [inline python/paddle/distributed/collective.py]: group (Group, optional): The group instance return by new_group or None for global default group.Default: None.
下同

**Review by TCChenlong — DISMISSED** [review]
LGTM
TODO：Fix Docs

**Review by wangxicoding — DISMISSED** [review]

**Review by ForFishes — DISMISSED** [review]
LGTM

**Review by XieYunshen — DISMISSED** [review]

**Review by gongweibao — DISMISSED** [review]
LGTM

**Review by lanxianghit — DISMISSED** [review]

**Review by fuyinno4 — APPROVED** [review]

**Review by XieYunshen — APPROVED** [review]

**Review by lanxianghit — APPROVED** [review]

File: python/paddle/distributed/collective.py (hunk 3 above)
Comment (sandyhouse) [reply python/paddle/distributed/collective.py]: Done.

File: python/paddle/distributed/collective.py (hunk 2 above)
Comment (sandyhouse) [reply python/paddle/distributed/collective.py:1206]: Done.

**Review by TCChenlong — APPROVED** [review]
LGTM

**Pull request merged**